File: codes/dart/array_filter.dart
import "dart:convert";

void main() {
    dynamic jsonStringify(dynamic anything, { bool pretty = false, String indent = "    " }) {
		if (pretty == true) {
			final jsonEncoder = new JsonEncoder.withIndent(indent);
			return jsonEncoder.convert(anything);
		}
		final jsonEncoder = new JsonEncoder();
		return jsonEncoder.convert(anything).split(',').join(', ');
    }

    dynamic arrayFilterV1(dynamic callbackFunction, dynamic anArray) {
        // JavaScript-like Array.filter() function
        dynamic dataFiltered = <dynamic>[];
        for (final objectEntry in anArray.asMap().entries) {
            final arrayItemIndex = objectEntry.key;
            final arrayItem = objectEntry.value;
            dynamic isConditionMatch = callbackFunction(arrayItem, arrayItemIndex, anArray);
            if (isConditionMatch == true) dataFiltered.add(arrayItem);
        }
        return dataFiltered;
    }

    dynamic arrayFilterV2(dynamic callbackFunction, dynamic anArray) {
        // JavaScript-like Array.filter() function
        dynamic dataFiltered = <dynamic>[];
        for (final objectEntry in anArray.asMap().entries) {
            final arrayItemIndex = objectEntry.key;
            final arrayItem = objectEntry.value;
            if (callbackFunction(arrayItem, arrayItemIndex, anArray) == true) dataFiltered.add(arrayItem);
        }
        return dataFiltered;
    }

    print("\n// JavaScript-like Array.filter() in Dart List");

    final numbers = <dynamic>[12, 34, 27, 23, 65, 93, 36, 87, 4, 254];
    print("numbers: ${jsonStringify(numbers)}");

    dynamic numbersEven;
    dynamic numbersOdd;

    print("// using JavaScript-like Array.filter() function \"arrayFilterV1\"");

    numbersEven = arrayFilterV1((dynamic number, dynamic _, dynamic __) => ((number % 2) == 0), numbers);
    print("even numbers only: ${jsonStringify(numbersEven)}");
    // even numbers only: [12, 34, 36, 4, 254]

    numbersOdd = arrayFilterV1((dynamic number, dynamic _, dynamic __) => ((number % 2) != 0), numbers);
    print("odd numbers only: ${jsonStringify(numbersOdd)}");
    // odd numbers only: [27, 23, 65, 93, 87]

    print("// using JavaScript-like Array.filter() function \"arrayFilterV2\"");

    numbersEven = arrayFilterV2((dynamic number, dynamic _, dynamic __) => ((number % 2) == 0), numbers);
    print("even numbers only: ${jsonStringify(numbersEven)}");
    // even numbers only: [12, 34, 36, 4, 254]

    numbersOdd = arrayFilterV2((dynamic number, dynamic _, dynamic __) => ((number % 2) != 0), numbers);
    print("odd numbers only: ${jsonStringify(numbersOdd)}");
    // odd numbers only: [27, 23, 65, 93, 87]

    print("// using Dart Array.filter() built-in method \"Array.where().toList()\"");

    numbersEven = numbers.where((dynamic number) => ((number % 2) == 0)).toList();
    print("even numbers only: ${jsonStringify(numbersEven)}");
    // even numbers only: [12, 34, 36, 4, 254]

    numbersOdd = numbers.where((dynamic number) => ((number % 2) != 0)).toList();
    print("odd numbers only: ${jsonStringify(numbersOdd)}");
    // odd numbers only: [27, 23, 65, 93, 87]

    print("\n// JavaScript-like Array.filter() in Dart List of Maps");

    final products = <dynamic>[
        <String, dynamic>{
            "code": "pasta",
            "price": 321
        },
        <String, dynamic>{
            "code": "bubble_gum",
            "price": 233
        },
        <String, dynamic>{
            "code": "potato_chips",
            "price": 5
        },
        <String, dynamic>{
            "code": "towel",
            "price": 499
        }
    ];
    print("products: ${jsonStringify(products, pretty: true)}");

    dynamic productsBelow100;
    dynamic productsAbove100;

    print("// using JavaScript-like Array.filter() function \"arrayFilterV1\"");

    productsBelow100 = arrayFilterV1((dynamic product, dynamic _, dynamic __) => (product["price"] <= 100), products);
    print("products with price <= 100 only: ${jsonStringify(productsBelow100, pretty: true)}");
    // products with price <= 100 only: [
    //     {
    //         "code": "potato_chips",
    //         "price": 5
    //     }
    // ]

    productsAbove100 = arrayFilterV1((dynamic product, dynamic _, dynamic __) => (product["price"] > 100), products);
    print("products with price > 100 only: ${jsonStringify(productsAbove100, pretty: true)}");
    // products with price > 100 only: [
    //     {
    //         "code": "pasta",
    //         "price": 321
    //     },
    //     {
    //         "code": "bubble_gum",
    //         "price": 233
    //     },
    //     {
    //         "code": "towel",
    //         "price": 499
    //     }
    // ]

    print("// using JavaScript-like Array.filter() function \"arrayFilterV2\"");

    productsBelow100 = arrayFilterV2((dynamic product, dynamic _, dynamic __) => (product["price"] <= 100), products);
    print("products with price <= 100 only: ${jsonStringify(productsBelow100, pretty: true)}");
    // products with price <= 100 only: [
    //     {
    //         "code": "potato_chips",
    //         "price": 5
    //     }
    // ]

    productsAbove100 = arrayFilterV2((dynamic product, dynamic _, dynamic __) => (product["price"] > 100), products);
    print("products with price > 100 only: ${jsonStringify(productsAbove100, pretty: true)}");
    // products with price > 100 only: [
    //     {
    //         "code": "pasta",
    //         "price": 321
    //     },
    //     {
    //         "code": "bubble_gum",
    //         "price": 233
    //     },
    //     {
    //         "code": "towel",
    //         "price": 499
    //     }
    // ]

    print("// using Dart Array.filter() built-in method \"Array.where().toList()\"");

    productsBelow100 = products.where((dynamic product) => (product["price"] <= 100)).toList();
    print("products with price <= 100 only: ${jsonStringify(productsBelow100, pretty: true)}");
    // products with price <= 100 only: [
    //     {
    //         "code": "potato_chips",
    //         "price": 5
    //     }
    // ]

    productsAbove100 = products.where((dynamic product) => (product["price"] > 100)).toList();
    print("products with price > 100 only: ${jsonStringify(productsAbove100, pretty: true)}");
    // products with price > 100 only: [
    //     {
    //         "code": "pasta",
    //         "price": 321
    //     },
    //     {
    //         "code": "bubble_gum",
    //         "price": 233
    //     },
    //     {
    //         "code": "towel",
    //         "price": 499
    //     }
    // ]
}

File: codes/csharp/array_filter.cs
using System; // Console, Func<>, Action<>, Action
using System.Linq; // Enumerable
using System.Collections.Generic; // Dictionary<>, List<>, KeyValuePair<>, IEnumerable<>

class Program {
    static void Main(string[] Args) {
        dynamic IsNumeric = (Func<dynamic, bool>)((dynamic Anything) => (Anything is sbyte || Anything is byte || Anything is short || Anything is ushort || Anything is int || Anything is uint || Anything is long || Anything is ulong || Anything is float || Anything is double || Anything is decimal));

        string JsonStringify(dynamic Anything, bool Pretty = false, string Indent = "    ") {
            int IndentLevel = 0;
            string JsonStringifyInner(dynamic AnythingInner, string IndentInner) {
                if (AnythingInner == null) return "null";
                if (AnythingInner is string) return "\"" + (string)AnythingInner + "\"";
                if ((IsNumeric(AnythingInner) == true) || AnythingInner is bool) return AnythingInner.ToString().Replace(",", ".");
                if (AnythingInner is List<dynamic>) {
                    IndentLevel += 1;
                    string Result = ((Pretty == true) ? ("[" + Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel))) : "[");
                    int ArrayItemIndex = 0;
                    foreach (dynamic ArrayItem in (List<dynamic>)AnythingInner) {
                        Result += JsonStringifyInner(ArrayItem, IndentInner);
                        if ((ArrayItemIndex + 1) != ((List<dynamic>)AnythingInner).Count) Result += ((Pretty == true) ? ("," + Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel))) : ", ");
                        ArrayItemIndex += 1;
                    }
                    IndentLevel -= 1;
                    Result += ((Pretty == true) ? (Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) + "]") : "]");
                    return Result;
                }
                if (AnythingInner is Dictionary<string, dynamic>) {
                    IndentLevel += 1;
                    string Result = ((Pretty == true) ? ("{" + Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel))) : "{");
                    int ObjectIterationIndex = 0;
                    foreach (KeyValuePair<string, dynamic> ObjectEntry in (Dictionary<string, dynamic>)AnythingInner) {
                        string ObjectKey = ObjectEntry.Key;
                        dynamic ObjectValue = ObjectEntry.Value;
                        Result += "\"" + ObjectKey + "\": " + JsonStringifyInner(ObjectValue, IndentInner);
                        if ((ObjectIterationIndex + 1) != ((Dictionary<string, dynamic>)AnythingInner).Count) Result += ((Pretty == true) ? ("," + Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel))) : ", ");
                        ObjectIterationIndex += 1;
                    }
                    IndentLevel -= 1;
                    Result += ((Pretty == true) ? (Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) + "}") : "}");
                    return Result;
                }
                return "null";
            };
            return JsonStringifyInner(Anything, Indent);
        }

        List<dynamic> ArrayFilterV1(Func<dynamic, dynamic, List<dynamic>, bool> CallbackFunction, List<dynamic> AnArray) {
            // JavaScript-like Array.filter() function
            List<dynamic> DataFiltered = new List<dynamic>();
            int ArrayItemIndex = 0;
            foreach (dynamic ArrayItem in AnArray) {
                bool IsConditionMatch = (bool)CallbackFunction(ArrayItem, ArrayItemIndex, AnArray);
                if (IsConditionMatch == true) DataFiltered.Add(ArrayItem);
                ArrayItemIndex += 1;
            }
            return DataFiltered;
        }

        List<dynamic> ArrayFilterV2(Func<dynamic, dynamic, List<dynamic>, bool> CallbackFunction, List<dynamic> AnArray) {
            // JavaScript-like Array.filter() function
            List<dynamic> DataFiltered = new List<dynamic>();
            int ArrayItemIndex = 0;
            foreach (dynamic ArrayItem in AnArray) {
                if ((bool)CallbackFunction(ArrayItem, ArrayItemIndex, AnArray) == true) DataFiltered.Add(ArrayItem);
                ArrayItemIndex += 1;
            }
            return DataFiltered;
        }

        Console.WriteLine("\n// JavaScript-like Array.filter() in C# List");

        dynamic Numbers = new List<dynamic>() {12, 34, 27, 23, 65, 93, 36, 87, 4, 254};
        Console.WriteLine($"Numbers: {JsonStringify(Numbers)}");

        dynamic NumbersEven;
        dynamic NumbersOdd;

        Console.WriteLine("// using JavaScript-like Array.filter() function \"ArrayFilterV1\"");

        NumbersEven = ArrayFilterV1((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Number, dynamic ArrayItemIndex, List<dynamic> AnArray) => ((Number % 2) == 0)), Numbers);
        Console.WriteLine($"even numbers only: {JsonStringify(NumbersEven)}");
        // even numbers only: [12, 34, 36, 4, 254]

        NumbersOdd = ArrayFilterV1((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Number, dynamic ArrayItemIndex, List<dynamic> AnArray) => ((Number % 2) != 0)), Numbers);
        Console.WriteLine($"odd numbers only: {JsonStringify(NumbersOdd)}");
        // odd numbers only: [27, 23, 65, 93, 87]

        Console.WriteLine("// using JavaScript-like Array.filter() function \"ArrayFilterV2\"");

        NumbersEven = ArrayFilterV2((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Number, dynamic ArrayItemIndex, List<dynamic> AnArray) => ((Number % 2) == 0)), Numbers);
        Console.WriteLine($"even numbers only: {JsonStringify(NumbersEven)}");
        // even numbers only: [12, 34, 36, 4, 254]

        NumbersOdd = ArrayFilterV2((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Number, dynamic ArrayItemIndex, List<dynamic> AnArray) => ((Number % 2) != 0)), Numbers);
        Console.WriteLine($"odd numbers only: {JsonStringify(NumbersOdd)}");
        // odd numbers only: [27, 23, 65, 93, 87]

        Console.WriteLine("// using C# Array.filter() built-in method \"IEnumerable.Where().ToList()\"");

        NumbersEven = ((IEnumerable<dynamic>)Numbers).Where((Func<dynamic, bool>)((dynamic Number) => ((Number % 2) == 0))).ToList();
        Console.WriteLine($"even numbers only: {JsonStringify(NumbersEven)}");
        // even numbers only: [12, 34, 36, 4, 254]

        NumbersOdd = ((IEnumerable<dynamic>)Numbers).Where((Func<dynamic, bool>)((dynamic Number) => ((Number % 2) != 0))).ToList();
        Console.WriteLine($"odd numbers only: {JsonStringify(NumbersOdd)}");
        // odd numbers only: [27, 23, 65, 93, 87]

        Console.WriteLine("\n// JavaScript-like Array.filter() in C# List of Dictionaries");

        dynamic Products = new List<dynamic>() {
            new Dictionary<string, dynamic>() {
                {"code", "pasta"},
                {"price", 321}
            },
            new Dictionary<string, dynamic>() {
                {"code", "bubble_gum"},
                {"price", 233}
            },
            new Dictionary<string, dynamic>() {
                {"code", "potato_chips"},
                {"price", 5}
            },
            new Dictionary<string, dynamic>() {
                {"code", "towel"},
                {"price", 499}
            },
        };
        Console.WriteLine($"Products: {JsonStringify(Products, Pretty: true)}");

        dynamic ProductsBelow100;
        dynamic ProductsAbove100;

        Console.WriteLine("// using JavaScript-like Array.filter() function \"ArrayFilterV1\"");

        ProductsBelow100 = ArrayFilterV1((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Product, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Product["price"] <= 100)), Products);
        Console.WriteLine($"products with price <= 100 only: {JsonStringify(ProductsBelow100, Pretty: true)}");
        // products with price <= 100 only: [
        //     {
        //         "code": "potato_chips",
        //         "price": 5
        //     }
        // ]

        ProductsAbove100 = ArrayFilterV1((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Product, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Product["price"] > 100)), Products);
        Console.WriteLine($"products with price > 100 only: {JsonStringify(ProductsAbove100, Pretty: true)}");
        // products with price > 100 only: [
        //     {
        //         "code": "pasta",
        //         "price": 321
        //     },
        //     {
        //         "code": "bubble_gum",
        //         "price": 233
        //     },
        //     {
        //         "code": "towel",
        //         "price": 499
        //     }
        // ]

        Console.WriteLine("// using JavaScript-like Array.filter() function \"ArrayFilterV2\"");

        ProductsBelow100 = ArrayFilterV2((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Product, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Product["price"] <= 100)), Products);
        Console.WriteLine($"products with price <= 100 only: {JsonStringify(ProductsBelow100, Pretty: true)}");
        // products with price <= 100 only: [
        //     {
        //         "code": "potato_chips",
        //         "price": 5
        //     }
        // ]

        ProductsAbove100 = ArrayFilterV2((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Product, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Product["price"] > 100)), Products);
        Console.WriteLine($"products with price > 100 only: {JsonStringify(ProductsAbove100, Pretty: true)}");
        // products with price > 100 only: [
        //     {
        //         "code": "pasta",
        //         "price": 321
        //     },
        //     {
        //         "code": "bubble_gum",
        //         "price": 233
        //     },
        //     {
        //         "code": "towel",
        //         "price": 499
        //     }
        // ]

        Console.WriteLine("// using C# Array.filter() built-in method \"IEnumerable.Where().ToList()\"");

        ProductsBelow100 = ((IEnumerable<dynamic>)Products).Where((Func<dynamic, bool>)((dynamic Product) => (Product["price"] <= 100))).ToList();
        Console.WriteLine($"products with price <= 100 only: {JsonStringify(ProductsBelow100, Pretty: true)}");
        // products with price <= 100 only: [
        //     {
        //         "code": "potato_chips",
        //         "price": 5
        //     }
        // ]

        ProductsAbove100 = ((IEnumerable<dynamic>)Products).Where((Func<dynamic, bool>)((dynamic Product) => (Product["price"] > 100))).ToList();
        Console.WriteLine($"products with price > 100 only: {JsonStringify(ProductsAbove100, Pretty: true)}");
        // products with price > 100 only: [
        //     {
        //         "code": "pasta",
        //         "price": 321
        //     },
        //     {
        //         "code": "bubble_gum",
        //         "price": 233
        //     },
        //     {
        //         "code": "towel",
        //         "price": 499
        //     }
        // ]
    }
}
